File: src/1799. Maximize Score After N Operations/main.go
package main

func gcd(a, b int) int {
	if b == 0 {
		return a
	}
	return gcd(b, a%b)
}

func maxScore(nums []int) int {
	dp := make([]int, 1<<len(nums))
	for i := 0; i < len(dp); i++ {
		dp[i] = -1
	}

	return dfs(nums, dp, 0, 0)
}

func dfs(nums []int, dp []int, mask, pairsPicked int) int {
	if 2*pairsPicked == len(nums) {
		return 0
	}

	if dp[mask] != -1 {
		return dp[mask]
	}

	maxScore := 0

	for i := 0; i < len(nums); i++ {
		for j := i + 1; j < len(nums); j++ {
			if (((mask >> i) & 1) == 1) || (((mask >> j) & 1) == 1) {
				continue
			}

			newMask := mask | (1 << i) | (1 << j)

			currScore := (pairsPicked + 1) * gcd(nums[i], nums[j])
			remainingScore := dfs(nums, dp, newMask, pairsPicked+1)

			if currScore+remainingScore > maxScore {
				maxScore = currScore + remainingScore
			}
		}
	}

	dp[mask] = maxScore

	return maxScore
}

File: src/1799. Maximize Score After N Operations/main.rs
impl Solution {
    pub fn max_score(nums: Vec<i32>) -> i32 {
        let mut dp = vec![-1; 1 << nums.len()];
        Self::eval(&nums, 0, &mut dp)
    }

    fn gcd(a: i32, b: i32) -> i32 {
        if a < b { return Self::gcd(b, a); }
        if a % b == 0 { return b }
        Self::gcd(b, a % b)
    } 

    fn eval(nums: &Vec<i32>, mask: usize, dp: &mut Vec<i32>) -> i32 {
        if dp[mask] != -1 { return dp[mask] }

        let mut k = nums.len();
        for i in 0 .. nums.len() {
            if mask & 1 << i != 0 { k -= 1; }
        }
        k >>= 1;
        if k == 0 { return 0 } 

        for i in 0 .. nums.len() {
            for j in i + 1 .. nums.len() {
                if mask & 1 << i != 0 || mask & 1 << j != 0 { continue }
                let mask_new = mask | 1 << i | 1 << j;
                dp[mask] = dp[mask].max(k as i32 * Self::gcd(nums[i], nums[j]) + Self::eval(nums, mask_new, dp));
            }
        }
        
        dp[mask]
    }
}